core/src/main/java/org/infinispan/container/offheap/OffHeapDataContainer.java

```diff
@@ -327,30 +327,40 @@ protected void entryRetrieved(long entryAddress) {
     * key <b>must</b> be held before calling this method.
     * @param bucketHeadAddress the starting address of the address hash
     * @param key the key of the entry
+    * @param requireReturn whether this method is forced to return the entry removed (optimizations can be done if
+    *                      the entry is not needed)
     */
-   protected InternalCacheEntry<WrappedBytes, WrappedBytes> performRemove(long bucketHeadAddress, Object key) {
+   protected InternalCacheEntry<WrappedBytes, WrappedBytes> performRemove(long bucketHeadAddress, Object key,
+         boolean requireReturn) {
       WrappedByteArray wba = toWrapper(key);
       long prevAddress = 0;
       // We only use the head pointer for the first iteration
       long address = bucketHeadAddress;
+      InternalCacheEntry<WrappedBytes, WrappedBytes> ice = null;
 
       while (address != 0) {
          long nextAddress = offHeapEntryFactory.getNext(address);
-         InternalCacheEntry<WrappedBytes, WrappedBytes> ice = offHeapEntryFactory.fromMemory(address);
-         if (ice.getKey().equals(wba)) {
+         boolean removeThisAddress;
+         if (requireReturn) {
+            ice = offHeapEntryFactory.fromMemory(address);
+            removeThisAddress = ice.getKey().equals(wba);
+         } else {
+            removeThisAddress = offHeapEntryFactory.equalsKey(address, wba);
+         }
+         if (removeThisAddress) {
             entryRemoved(address);
             if (prevAddress != 0) {
                offHeapEntryFactory.setNext(prevAddress, nextAddress);
             } else {
                memoryLookup.putMemoryAddress(key, nextAddress);
             }
             size.decrementAndGet();
-            return ice;
+            break;
```

Comment:
Doesn't this mean if none of the entries in the bucket match the key, the method returns the last entry in the bucket?

Reply:
When `requireReturn` is true, yeah. Good catch. I can fix this by adding an else and assigning to null or if we change the other comment it would be fine.